consensus/polybft/validator_set.go

```diff
@@ -18,46 +86,430 @@ type ValidatorSet interface {
+	}
+
+	return proposer.Metadata.Address, nil
+}
+
+// GetProposer returns the current proposer. If the validator set is empty, nil
+// is returned.
+func (v *validatorSet) GetProposer() (*ValidatorAccount, error) {
+	if v.IsNilOrEmpty() {
+		return nil, fmt.Errorf("validators cannot be nil or empty")
+	}
+
+	if v.proposer == nil {
+		proposer, err := v.findProposer()
+
+		if err != nil {
+			return nil, fmt.Errorf("cannot find proposer: %w", err)
+		}
+
+		v.proposer = proposer
+	}
+
+	return v.proposer.Copy(), nil
+}
+
+func (v *validatorSet) findProposer() (*ValidatorAccount, error) {
+	var (
+		proposer *ValidatorAccount
+		err      error
+	)
+
+	for _, val := range v.validators {
+		if proposer == nil || !bytes.Equal(val.Metadata.Address.Bytes(), proposer.Metadata.Address.Bytes()) {
+			proposer, err = proposer.CompareProposerPriority(val)
+			if err != nil {
+				return nil, fmt.Errorf("cannot compare proposer priority: %w", err)
+			}
+		}
+	}
+
+	return proposer, nil
+}
+
+func (v *validatorSet) Includes(address types.Address) bool {
+	for _, validator := range v.validators {
+		if validator.Metadata.Address == address {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v *validatorSet) Len() int {
+	return len(v.validators)
+}
+
+// Compute the difference between the max and min ProposerPriority of that set.
+func computeMaxMinPriorityDiff(v *validatorSet) int64 {
+	max := int64(math.MinInt64)
+	min := int64(math.MaxInt64)
+
+	for _, v := range v.validators {
+		if v.ProposerPriority < min {
+			min = v.ProposerPriority
+		}
+
+		if v.ProposerPriority > max {
+			max = v.ProposerPriority
+		}
+	}
+
+	diff := max - min
+
+	if diff < 0 {
+		return -1 * diff
+	}
+
+	return diff
+}
+
+// Copy each validator into a new ValidatorSet.
+func (v *validatorSet) Copy() *validatorSet {
+	return &validatorSet{
+		validators:       validatorListCopy(v.validators),
+		proposer:         v.proposer,
+		totalVotingPower: v.totalVotingPower,
+	}
+}
+
+// Makes a copy of the validator list.
+func validatorListCopy(valList []*ValidatorAccount) []*ValidatorAccount {
+	valCopy := make([]*ValidatorAccount, len(valList))
+	for i, val := range valList {
+		valCopy[i] = val.Copy()
+	}
+
+	return valCopy
+}
+
+func safeAdd(a, b int64) (int64, bool) {
+	if b > 0 && a > math.MaxInt64-b {
+		return -1, true
+	} else if b < 0 && a < math.MinInt64-b {
+		return -1, true
+	}
+
+	return a + b, false
+}
+
+func safeSub(a, b int64) (int64, bool) {
+	if b > 0 && a < math.MinInt64+b {
+		return -1, true
+	} else if b < 0 && a > math.MaxInt64+b {
+		return -1, true
+	}
+
+	return a - b, false
+}
+
+func safeAddClip(a, b int64) int64 {
+	c, overflow := safeAdd(a, b)
+	if overflow {
+		if b < 0 {
+			return math.MinInt64
+		}
+
+		return math.MaxInt64
+	}
+
+	return c
+}
+
+func safeSubClip(a, b int64) int64 {
+	c, overflow := safeSub(a, b)
+	if overflow {
+		if b > 0 {
+			return math.MinInt64
+		}
+
+		return math.MaxInt64
+	}
+
+	return c
 }
```

Comment:
It'd be nice if we would extract this away from `validator_set` and put it to some `math.go` in the `helper/common` package?

Reply:
This is not ready for review.